src/hawkes/hawkes_process.jl

```diff
@@ -0,0 +1,169 @@
+struct HawkesProcess{T<:Real} <: AbstractPointProcess
+    μ::T
+    α::T
+    ω::T
+
+    function HawkesProcess(μ::T1, α::T2, ω::T3) where {T1,T2,T3}
+        any((μ, α, ω) .< 0) &&
+            throw(DomainError((μ, α, ω), "All parameters must be non-negative."))
+        T = promote_type(T1, T2, T3)
+        (μ_T, α_T, ω_T) = convert.(T, (μ, α, ω))
+        new{T}(μ_T, α_T, ω_T)
+    end
+end
+
+function Base.rand(rng::AbstractRNG, hp::HawkesProcess, tmin, tmax)
+    sim = simulate_poisson_times(rng, hp.μ, tmin, tmax) # Simulate Poisson process with base rate
+    sim_desc = generate_descendants(rng, sim, tmax, hp.α, hp.ω) # Recursively generates descendants from first events
+    append!(sim, sim_desc)
+    sort!(sim)
+    return History(sim, fill(nothing, length(sim)), tmin, tmax)
+end
+
+"""
+    StatsAPI.fit(rng, ::HawkesProcess, h::History; step_tol::Float64 = 1e-3, max_iter::Int = 1000)
+
+Expectation-Maximization algorithm from [E. Lewis, G. Mohler (2011)](https://api.semanticscholar.org/CorpusID:2105771)).
+The relevant calculations are in page 4, equations 6-13.
+
+Let (t_1 < ... < t_N) be the event times over the interval [0, T). We use the immigrant-descendant representation,
+where immigrants arrive at a constant base rate μ and each each arrival may generate descendants following the
+activation function α exp(-ω(t - t_i)).
+
+The algorithm consists in the following steps:
+1. Start with some initial guess for the parameters μ, ψ, and ω. ψ = α ω is the branching factor.
+2. Calculate λ(t_i; μ, ψ, ω) (`lambda_ts` in the code) using the procedure in [Ozaki (1979)](https://doi.org/10.1007/bf02480272).
+3. For each t_i and each j < i, calculate D_{ij} = P(t_i is a descendant of t_j).
+        D_{ij} = ψ ω exp(-ω(t_i - t_j)) / λ(t_i; μ, ψ, ω)
+    Define D = sum_{j < i} D_{ij} (expected number of descendants) and div = sum_{j < i} (t_i - t_j) D_ij. 
+4. Update the parameters as
+        μ = (N - D) / T
+        ψ = D / N
+        ω = D / div
+5. If convergence criterion is met, return updated parameters, otherwise, back to step 2.
+
+Notice that in the implementation the process is normalized so the average inter-event time is equal to 1 and, 
+therefore, the interval of the process is transformed from T to N. Also, in equation (8) in the paper,
+sum_{i=1}^N p_{ii} = sum_{i=1}^N (1 - sum{j<i} D_{ij}) = N - D.
+"""
+function StatsAPI.fit(
+    rng::AbstractRNG,
+    ::Type{HawkesProcess},
+    h::History;
+    step_tol::Float64=1e-4,
+    max_iter::Int=1000,
+)
+    N = nb_events(h)
+    N == 0 && return HawkesProcess(0.0, 0.0, 0.0)
+    T = duration(h)
+    norm_ts = h.times .* (N / T) # Average inter-event time equal to 1. Numerical stability
+    n_iters = 0
+    step = step_tol + 1.0
```

Comment:
Whenever you use floating point numbers like `1.0`, you're forcing the entire computation to happen in `Float64`, even if the user provided `Float32` everywhere. This could be replaced by `one(T)` for the right choice of `T`.

What is the right choice of `T`? Probably the result of the promotion between the time type of the `History` and the parameter type of the `HawkesProcess`